File: platform/sdk/dotnet/tests/LayerX.Sdk.Tests/ReceiptFixtureTests.cs
using System.Runtime.CompilerServices;
using System.Text.Json;
using LayerX.Sdk;
using Xunit;

namespace LayerX.Sdk.Tests;

public sealed class ReceiptFixtureTests
{
    private const string ProgramOutcomeV3 = "505247330100000000000100010000000700000001000000000000000b000000000000000c000000000000000d000000000000000e00000001000000000000000f0000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000020000000000000003000000000000000400000000000000050000000000000006000000000000000700000020000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000201111111111111111111111111111111111111111111111111111111111111111000000202222222222222222222222222222222222222222222222222222222222222222000000200000000000000000000000000000000000000000000000000000000000000000";

    [Fact]
    public void NativeSignedBinding()
    {
        using var document = JsonDocument.Parse(File.ReadAllText(FixturePath("native-program-call-v3.json")));
        var root = document.RootElement;
        var payload = HexField(root, "payload_hex"); var nativeCall = NativeProgramCall.Decode(payload);
        Assert.Equal(payload, nativeCall.Encode()); var signed = HexField(root, "signed_activity_hex");
        var method = typeof(ProgramsClient).GetMethod("DecodeSignedCall", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic)!;
        method.Invoke(null, new object[] { new ProgramCall(nativeCall, new ProtocolAmount("1000"), signed) });
        Assert.Throws<System.Reflection.TargetInvocationException>(() => method.Invoke(null, new object[] { new ProgramCall(nativeCall, new ProtocolAmount("999"), signed) }));
        Assert.Throws<System.Reflection.TargetInvocationException>(() => method.Invoke(null, new object[] { new ProgramCall(nativeCall with { ResponseCapacity = 17 }, new ProtocolAmount("1000"), signed) }));
        for (var length = 0; length < payload.Length; length++) Assert.Throws<ArgumentException>(() => NativeProgramCall.Decode(payload[..length]));
    }

    [Fact]
    public async Task ExplicitProtocolThree()
    {
        foreach (var name in new[] { "receipt-positive-v3.json", "receipt-programs-positive-v3.json" })
        {
            var fixture = LoadFixture(name);
            await Assert.ThrowsAsync<PlatformSdkException>(async () => await LocalVerifier.VerifyReceiptAsync(fixture.CanonicalReceipt, fixture.Batch));
            var verified = await LocalVerifier.VerifyReceiptAsync(fixture.CanonicalReceipt, fixture.Batch, protocolVersion: 3);
            Assert.Equal((ushort)3, verified.Receipt.ProtocolVersion);
            Assert.Equal(HexField(fixture.Expected, "receipt_digest_hex"), verified.ReceiptDigest);
            var corrupted = (byte[])fixture.CanonicalReceipt.Clone(); corrupted[^1] ^= 1;
            await Assert.ThrowsAsync<PlatformSdkException>(async () => await LocalVerifier.VerifyReceiptAsync(corrupted, fixture.Batch, protocolVersion: 3));
        }
    }

    [Fact]
    public void ProgramOutcomeV3VectorDecodes()
    {
        var outcome = LocalVerifier.DecodeProgramReceiptOutcome(Convert.FromHexString(ProgramOutcomeV3), 1);
        Assert.Equal((byte)3, outcome.EncodingVersion);
        Assert.Equal((ushort)1, outcome.AbiVersion);
        Assert.Equal(new UInt128Value(0, 16), outcome.FeeUnits);
        Assert.Equal(Enumerable.Repeat((byte)0x11, 32).ToArray(), outcome.CallGraphRoot);
        Assert.Equal(Enumerable.Repeat((byte)0x22, 32).ToArray(), outcome.TerminalPayloadRoot);
    }
    private sealed record Fixture(
        byte[] CanonicalReceipt,
        AuthorizedReceiptBatch Batch,
        JsonElement Expected,
        JsonElement AuthorizedBatch);

    private static string RepoRoot([CallerFilePath] string sourcePath = "")
        => Path.GetFullPath(Path.Combine(
            Path.GetDirectoryName(sourcePath) ?? throw new InvalidOperationException("no source dir"),
            "..", "..", "..", "..", ".."));

    private static byte[] HexField(JsonElement element, string key)
        => Convert.FromHexString(element.GetProperty(key).GetString()
            ?? throw new InvalidOperationException($"missing {key}"));

    private static UInt128Value U128Field(JsonElement element, string key)
        => new(0, ulong.Parse(element.GetProperty(key).GetString()
            ?? throw new InvalidOperationException($"missing {key}")));

    private static string FixturePath(string name) => Path.Combine(
        RepoRoot(), "platform", "sdk", "conformance", "fixtures", name);

    private static Fixture LoadFixture(string name = "receipt-positive-v2.json")
    {
        var path = FixturePath(name);
        using var document = JsonDocument.Parse(File.ReadAllText(path));
        var root = document.RootElement.Clone();
        var batch = root.GetProperty("authorized_batch");
        return new Fixture(
            HexField(root, "canonical_receipt_hex"),
            new AuthorizedReceiptBatch(
                HexField(batch, "batch_id_hex"),
                HexField(batch, "asset_hex"),
                HexField(batch, "previous_state_root_hex"),
                HexField(batch, "resulting_state_root_hex"),
                HexField(batch, "sequencer_public_key_hex")),
            root.GetProperty("expected"),
            batch);
    }

    [Fact]
    public async Task CoreFixtureReceiptVerifiesPositively()
    {
        var fixture = LoadFixture();
        var expected = fixture.Expected;
        var verified = await LocalVerifier.VerifyReceiptAsync(fixture.CanonicalReceipt, fixture.Batch);
        Assert.Equal(expected.GetProperty("level").GetString(), verified.Level);
        Assert.Equal(fixture.CanonicalReceipt, verified.CanonicalBytes);
        Assert.Equal(HexField(expected, "receipt_digest_hex"), verified.ReceiptDigest);
        var receipt = verified.Receipt;
        Assert.Equal(expected.GetProperty("result_code").GetInt32(), receipt.ResultCode);
        Assert.Equal(expected.GetProperty("protocol_version").GetUInt16(), receipt.ProtocolVersion);
        Assert.Equal(expected.GetProperty("operation").GetByte(), receipt.Operation);
        Assert.Equal(expected.GetProperty("module_id").GetUInt16(), receipt.ModuleId);
        Assert.Equal(expected.GetProperty("global_sequence").GetUInt64(), receipt.GlobalSequence);
        Assert.Equal(expected.GetProperty("timestamp_ms").GetUInt64(), receipt.Timestamp);
        Assert.Equal(U128Field(expected, "amount"), receipt.Amount);
        Assert.Equal(U128Field(expected, "fee_charged"), receipt.FeeCharged);
        Assert.Equal(U128Field(expected, "from_balance_before"), receipt.FromBalanceBefore);
        Assert.Equal(U128Field(expected, "from_balance_after"), receipt.FromBalanceAfter);
        Assert.Equal(U128Field(expected, "to_balance_before"), receipt.ToBalanceBefore);
        Assert.Equal(U128Field(expected, "to_balance_after"), receipt.ToBalanceAfter);
        Assert.Equal(HexField(expected, "activity_id_hex"), receipt.ActivityId);
        Assert.Equal(HexField(expected, "from_hex"), receipt.From);
        Assert.Equal(HexField(expected, "to_hex"), receipt.To);
        Assert.Equal(HexField(fixture.AuthorizedBatch, "batch_id_hex"), receipt.BatchId);
        Assert.Equal(HexField(fixture.AuthorizedBatch, "asset_hex"), receipt.Asset);
        Assert.Equal(
            HexField(fixture.AuthorizedBatch, "previous_state_root_hex"), receipt.PreviousStateRoot);
        Assert.Equal(
            HexField(fixture.AuthorizedBatch, "resulting_state_root_hex"), receipt.ResultingStateRoot);
    }

    [Fact]
    public async Task CoreFixtureReceiptByteFlipFails()
    {
        var fixture = LoadFixture();
        var mutated = (byte[])fixture.CanonicalReceipt.Clone();
        mutated[^1] ^= 0x01;
        await Assert.ThrowsAsync<PlatformSdkException>(async () =>
            await LocalVerifier.VerifyReceiptAsync(mutated, fixture.Batch));
    }

    [Fact]
    public async Task ProgramsReceiptPreservesOptionalOutcome()
    {
        var fixture = LoadFixture("receipt-programs-positive-v2.json");
        var verified = await LocalVerifier.VerifyReceiptAsync(fixture.CanonicalReceipt, fixture.Batch);
        var outcome = Assert.IsType<ProgramReceiptOutcome>(verified.Receipt.ProgramOutcome);
        Assert.Equal((byte)3, outcome.EncodingVersion);
        Assert.Equal((ushort)1, outcome.RuntimeVersion);
        Assert.Equal((ushort)1, outcome.AbiVersion);
        Assert.Equal(new UInt128Value(0, 2), outcome.OccupancyByteBatches);
        Assert.Equal(new UInt128Value(0, 7), outcome.OccupancyFeeUnits);
        Assert.Equal(fixture.Batch.Asset, outcome.OccupancyAssetId);
        Assert.Contains(outcome.OccupancyEvidenceDigest, value => value != 0);
        Assert.Contains(outcome.OccupancyTransferRoot, value => value != 0);
        Assert.Equal(new UInt128Value(0, 16), outcome.FeeUnits);
    }

    [Fact]
    public async Task RefusalVectorsExposeSharedTaxonomy()
    {
        using var document = JsonDocument.Parse(File.ReadAllText(
            FixturePath("receipt-refusals-v2.json")));
        var root = document.RootElement;
        var authority = root.GetProperty("authorized_batch");
        var batch = new AuthorizedReceiptBatch(
            HexField(authority, "batch_id_hex"),
            HexField(authority, "asset_hex"),
            HexField(authority, "previous_state_root_hex"),
            HexField(authority, "resulting_state_root_hex"),
            HexField(authority, "sequencer_public_key_hex"));
        foreach (var vector in root.GetProperty("vectors").EnumerateArray())
        {
            var failure = await Assert.ThrowsAsync<PlatformSdkException>(async () =>
                await LocalVerifier.VerifyReceiptAsync(
                    HexField(vector, "canonical_receipt_hex"), batch));
            Assert.NotNull(failure.ReceiptCheck);
            Assert.Equal(vector.GetProperty("expected_check").GetString(),
                failure.ReceiptCheck!.Value.MachineCode());
        }
    }
}

File: platform/sdk/swift/Tests/LayerXSDKTests/ReceiptFixtureTests.swift
import Foundation
import XCTest
@testable import LayerXSDK

final class ReceiptFixtureTests: XCTestCase {
    private static let programOutcomeV3 = "505247330100000000000100010000000700000001000000000000000b000000000000000c000000000000000d000000000000000e00000001000000000000000f0000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000020000000000000003000000000000000400000000000000050000000000000006000000000000000700000020000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000201111111111111111111111111111111111111111111111111111111111111111000000202222222222222222222222222222222222222222222222222222222222222222000000200000000000000000000000000000000000000000000000000000000000000000"

    func testProgramOutcomeV3Vector() throws {
        let hex = Self.programOutcomeV3
        let encoded = Data(stride(from: 0, to: hex.count, by: 2).map {
            UInt8(hex[hex.index(hex.startIndex, offsetBy: $0)..<hex.index(hex.startIndex, offsetBy: $0 + 2)], radix: 16)!
        })
        let outcome = try LocalVerifier.decodeProgramReceiptOutcome(encoded, protocolVersion: 1)
        XCTAssertEqual(outcome.encodingVersion, 3)
        XCTAssertEqual(outcome.abiVersion, 1)
        XCTAssertEqual(outcome.feeUnits, UInt128Value(high: 0, low: 16))
        XCTAssertEqual(outcome.callGraphRoot, Data(repeating: 0x11, count: 32))
        XCTAssertEqual(outcome.terminalPayloadRoot, Data(repeating: 0x22, count: 32))
    }
    private struct Fixture {
        let canonicalReceipt: Data
        let batch: AuthorizedReceiptBatch
        let expected: [String: Any]
        let authorizedBatch: [String: Any]
    }

    private func fixtureURL(_ name: String = "receipt-positive-v2.json") -> URL {
        var url = URL(fileURLWithPath: #filePath)
        for _ in 0..<6 { url.deleteLastPathComponent() }
        return url
            .appendingPathComponent("platform/sdk/conformance/fixtures")
            .appendingPathComponent(name)
    }

    private func hexData(_ value: String) throws -> Data {
        XCTAssertEqual(value.count % 2, 0, "odd hex length")
        var bytes = Data(capacity: value.count / 2)
        var index = value.startIndex
        while index < value.endIndex {
            let next = value.index(index, offsetBy: 2)
            let byte = try XCTUnwrap(UInt8(value[index..<next], radix: 16), "invalid hex byte")
            bytes.append(byte)
            index = next
        }
        return bytes
    }

    private func hexField(_ object: [String: Any], _ key: String) throws -> Data {
        try hexData(try XCTUnwrap(object[key] as? String, "missing \(key)"))
    }

    private func u128Field(_ object: [String: Any], _ key: String) throws -> UInt128Value {
        let text = try XCTUnwrap(object[key] as? String, "missing \(key)")
        return UInt128Value(high: 0, low: try XCTUnwrap(UInt64(text), "non-decimal \(key)"))
    }

    private func loadFixture(_ name: String = "receipt-positive-v2.json") throws -> Fixture {
        let raw = try Data(contentsOf: fixtureURL(name))
        let json = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: raw) as? [String: Any], "fixture is not an object")
        let authorizedBatch = try XCTUnwrap(
            json["authorized_batch"] as? [String: Any], "missing authorized_batch")
        let expected = try XCTUnwrap(json["expected"] as? [String: Any], "missing expected")
        let batch = AuthorizedReceiptBatch(
            batchID: try hexField(authorizedBatch, "batch_id_hex"),
            asset: try hexField(authorizedBatch, "asset_hex"),
            previousStateRoot: try hexField(authorizedBatch, "previous_state_root_hex"),
            resultingStateRoot: try hexField(authorizedBatch, "resulting_state_root_hex"),
            sequencerPublicKey: try hexField(authorizedBatch, "sequencer_public_key_hex"))
        return Fixture(
            canonicalReceipt: try hexField(json, "canonical_receipt_hex"),
            batch: batch,
            expected: expected,
            authorizedBatch: authorizedBatch)
    }

    func testCoreFixtureReceiptVerifiesPositively() async throws {
        let fixture = try loadFixture()
        let expected = fixture.expected
        let verified = try await LocalVerifier.verifyReceipt(
            fixture.canonicalReceipt, authorized: fixture.batch)
        XCTAssertEqual(verified.level, try XCTUnwrap(expected["level"] as? String))
        XCTAssertEqual(verified.canonicalBytes, fixture.canonicalReceipt)
        XCTAssertEqual(verified.receiptDigest, try hexField(expected, "receipt_digest_hex"))
        let receipt = verified.receipt
        XCTAssertEqual(
            Int64(receipt.resultCode),
            try XCTUnwrap(expected["result_code"] as? NSNumber).int64Value)
        XCTAssertEqual(
            UInt64(receipt.protocolVersion),
            try XCTUnwrap(expected["protocol_version"] as? NSNumber).uint64Value)
        XCTAssertEqual(
            UInt64(receipt.operation),
            try XCTUnwrap(expected["operation"] as? NSNumber).uint64Value)
        XCTAssertEqual(
            UInt64(receipt.moduleID),
            try XCTUnwrap(expected["module_id"] as? NSNumber).uint64Value)
        XCTAssertEqual(
            receipt.globalSequence,
            try XCTUnwrap(expected["global_sequence"] as? NSNumber).uint64Value)
        XCTAssertEqual(
            receipt.timestamp,
            try XCTUnwrap(expected["timestamp_ms"] as? NSNumber).uint64Value)
        XCTAssertEqual(receipt.amount, try u128Field(expected, "amount"))
        XCTAssertEqual(receipt.feeCharged, try u128Field(expected, "fee_charged"))
        XCTAssertEqual(receipt.fromBalanceBefore, try u128Field(expected, "from_balance_before"))
        XCTAssertEqual(receipt.fromBalanceAfter, try u128Field(expected, "from_balance_after"))
        XCTAssertEqual(receipt.toBalanceBefore, try u128Field(expected, "to_balance_before"))
        XCTAssertEqual(receipt.toBalanceAfter, try u128Field(expected, "to_balance_after"))
        XCTAssertEqual(receipt.activityID, try hexField(expected, "activity_id_hex"))
        XCTAssertEqual(receipt.from, try hexField(expected, "from_hex"))
        XCTAssertEqual(receipt.to, try hexField(expected, "to_hex"))
        XCTAssertEqual(receipt.batchID, try hexField(fixture.authorizedBatch, "batch_id_hex"))
        XCTAssertEqual(receipt.asset, try hexField(fixture.authorizedBatch, "asset_hex"))
        XCTAssertEqual(
            receipt.previousStateRoot,
            try hexField(fixture.authorizedBatch, "previous_state_root_hex"))
        XCTAssertEqual(
            receipt.resultingStateRoot,
            try hexField(fixture.authorizedBatch, "resulting_state_root_hex"))
    }

    func testCoreFixtureReceiptByteFlipFails() async throws {
        let fixture = try loadFixture()
        var mutated = fixture.canonicalReceipt
        mutated[mutated.count - 1] ^= 0x01
        do {
            _ = try await LocalVerifier.verifyReceipt(mutated, authorized: fixture.batch)
            XCTFail("mutated receipt verified; a flipped signature byte must fail")
        } catch {}
    }

    func testProgramsReceiptPreservesOptionalOutcome() async throws {
        let fixture = try loadFixture("receipt-programs-positive-v2.json")
        let verified = try await LocalVerifier.verifyReceipt(
            fixture.canonicalReceipt, authorized: fixture.batch)
        let outcome = try XCTUnwrap(verified.receipt.programOutcome)
        XCTAssertEqual(outcome.encodingVersion, 3)
        XCTAssertEqual(outcome.runtimeVersion, 1)
        XCTAssertEqual(outcome.abiVersion, 1)
        XCTAssertEqual(outcome.occupancyByteBatches, UInt128Value(high: 0, low: 2))
        XCTAssertEqual(outcome.occupancyFeeUnits, UInt128Value(high: 0, low: 7))
        XCTAssertEqual(outcome.occupancyAssetID, fixture.batch.asset)
        XCTAssertNotEqual(outcome.occupancyEvidenceDigest, Data(repeating: 0, count: 32))
        XCTAssertNotEqual(outcome.occupancyTransferRoot, Data(repeating: 0, count: 32))
        XCTAssertEqual(outcome.feeUnits, UInt128Value(high: 0, low: 16))
    }

    func testRefusalVectorsExposeSharedTaxonomy() async throws {
        let raw = try Data(contentsOf: fixtureURL("receipt-refusals-v2.json"))
        let json = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: raw) as? [String: Any])
        let authority = try XCTUnwrap(json["authorized_batch"] as? [String: Any])
        let batch = AuthorizedReceiptBatch(
            batchID: try hexField(authority, "batch_id_hex"),
            asset: try hexField(authority, "asset_hex"),
            previousStateRoot: try hexField(authority, "previous_state_root_hex"),
            resultingStateRoot: try hexField(authority, "resulting_state_root_hex"),
            sequencerPublicKey: try hexField(authority, "sequencer_public_key_hex"))
        let vectors = try XCTUnwrap(json["vectors"] as? [[String: Any]])
        for vector in vectors {
            let name = try XCTUnwrap(vector["name"] as? String)
            let expected = try XCTUnwrap(vector["expected_check"] as? String)
            do {
                _ = try await LocalVerifier.verifyReceipt(
                    try hexField(vector, "canonical_receipt_hex"), authorized: batch)
                XCTFail("\(name) verified")
            } catch let error as PlatformSDKError {
                XCTAssertEqual(error.receiptCheck?.rawValue, expected, name)
            }
        }
    }
}
